### PR TITLE
7175396: The text on label is not painted red for Nimbus LaF.

### PR DESCRIPTION
**Pull request by prsadhuk**
Label.foreground UIProperty is not honored by Nimbus L&F.
Added support for setting JLabel foreground color for Nimbus L&F

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-7175396](https://bugs.openjdk.org/browse/JDK-7175396): The text on label is not painted red for Nimbus LaF.

### Reviewers
 * [Harshitha Onkar](https://openjdk.org/census#honkar) (@honkar-jdk - Committer) ⚠️ Review applies to [b6d0a8bd](https://git.openjdk.org/jdk/pull/9900/files/b6d0a8bd76555032a5870ba667b661c155d64b7a)
 * [Tejesh R](https://openjdk.org/census#tr) (@TejeshR13 - Committer) ⚠️ Review applies to [e6de6d08](https://git.openjdk.org/jdk/pull/9900/files/e6de6d08a03e099a1b01365da41cdcdcebe031fe)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9900/head:pull/9900` \
`$ git checkout pull/9900`

Update a local copy of the PR: \
`$ git checkout pull/9900` \
`$ git pull https://git.openjdk.org/jdk pull/9900/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9900`

View PR using the GUI difftool: \
`$ git pr show -t 9900`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9900.diff">https://git.openjdk.org/jdk/pull/9900.diff</a>

</details>

### CODE REVIEW

**Review by honkar-jdk — COMMENTED** [review]
Copyright year for SynthLabelUI.java needs to be updated.
Applied the patch and tested, fix works.

File: test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        try {
+            UIManager.setLookAndFeel("javax.swing.plaf.nimbus.NimbusLookAndFeel");
+            Robot robot = new Robot();
+
+            SwingUtilities.invokeAndWait(() -> {
+                frame = new JFrame();
+                UIManager.getDefaults().put("Label.foreground",
+                                            java.awt.Color.red);
+                label =
+                    new JLabel("<html><body>Can You Read This?</body></html>");
```
[the remark this answers is not in the document]
Comment (prsadhuk) [reply test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java]: ok

**Review by TejeshR13 — APPROVED** [review]
Fix is working fine.

File: src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLabelUI.java [hunk 2]
```diff
@@ -214,6 +215,10 @@ protected void paint(SynthContext context, Graphics g) {
         } else {
             g.setColor(context.getStyle().getColor(context,
                     ColorType.TEXT_FOREGROUND));
+            Color col = UIManager.getColor("Label.foreground");
+            if (col != null) {
+                label.setForeground(col);
```
Comment (mrserb) [inline src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLabelUI.java:220]: Are you sure that the text color of the label should be changed on each repain?

**Comment by honkar-jdk** [comment]
Updated fix looks good.

**Review by honkar-jdk — APPROVED** [review]

File: src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLabelUI.java [hunk 3]
```diff
@@ -214,6 +217,9 @@ protected void paint(SynthContext context, Graphics g) {
         } else {
             g.setColor(context.getStyle().getColor(context,
                     ColorType.TEXT_FOREGROUND));
+            if (fgColor != null) {
+                label.setForeground(fgColor);
+            }
```
Comment (aivanov-jdk) [inline src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLabelUI.java:222]: Why is the color set to the `label` component during paint?

Shouldn't it be `g.setColor(fgColor)`? And the above code which uses the `context` should be skipped if `fgColor != null`.

File: test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    public static void main(String[] args) throws Exception {
+
+        try {
+            UIManager.setLookAndFeel("javax.swing.plaf.nimbus.NimbusLookAndFeel");
+            Robot robot = new Robot();
+
+            SwingUtilities.invokeAndWait(() -> {
+                frame = new JFrame();
+                UIManager.getDefaults().put("Label.foreground",
+                                            java.awt.Color.red);
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java:75]: You import `Color` class, there's no need to specify the full class name. You should use `RED` as constants are in upper case.

File: test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            Dimension d = frame.getSize();
+            System.out.println("width " + d.getWidth() +
+                                " height " + d.getHeight());
+            int y = p.y + d.height/2;
+
+            for (int x = p.x; x < p.x + d.width; x++) {
+                    System.out.println("color(" + x + "," + y + ")=" +
+                                        robot.getPixelColor(x, y));
+                    Color color = robot.getPixelColor(x, y);
+                    if (checkPixel(color)) {
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java:98]: I wonder if disabling text antialiasing would yield a simpler code in `checkPixel`? However, it cannot be disabled on macOS which still uses grayscale antialiasing. But for Windows and Linux, it should help. Or the test could be limited to Windows and Linux.

File: test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+            Point p = frame.getLocationOnScreen();
+            Dimension d = frame.getSize();
+            System.out.println("width " + d.getWidth() +
+                                " height " + d.getHeight());
+            int y = p.y + d.height/2;
+
+            for (int x = p.x; x < p.x + d.width; x++) {
+                    System.out.println("color(" + x + "," + y + ")=" +
+                                        robot.getPixelColor(x, y));
+                    Color color = robot.getPixelColor(x, y);
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java:97]:
```suggestion
                    Color color = robot.getPixelColor(x, y);
                    System.out.println("color(" + x + "," + y + ")=" +
                                        color);
```

File: test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+import javax.swing.JFrame;
+import javax.swing.JLabel;
+import javax.swing.SwingUtilities;
+import javax.swing.UIManager;
+
+public class TestNimbusLabel
+{
+    private static JLabel label;
+    private static JFrame frame;
+    private static boolean passed = false;
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java:48]: `passed` could be local variable in `main`.

`label` could also be local variable inside the lambda expression which creates the UI.

File: src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLabelUI.java (hunk 3 above)
Comment (prsadhuk) [reply src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLabelUI.java:222]: There are 2 ways JLabel can be used, one via html label and another normal label. Normal label text will work if we do `g.setColor(fgColor)` but html label will use [StyleSheet.paint](https://github.com/openjdk/jdk/blob/master/src/java.desktop/share/classes/javax/swing/text/html/StyleSheet.java#L1939) which does not take into account foreground Color (guess unless it's provided by html tags which is not the case here) so I need to explicitly set the foreground color of label

File: test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+                if (checkPixel(color)) {
+                    passed = true;
+                    break;
+                }
+            }
+            if (!passed) {
+                BufferedImage img =
+                    robot.createScreenCapture(new Rectangle(p.x, p.y,
+                            d.width,
+                            d.height));
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java:106]:
```suggestion
                    robot.createScreenCapture(new Rectangle(p, d));
```

File: test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        int green1 = color.getGreen();
+        int red2 = Color.red.getRed();
+        int blue2 = Color.red.getBlue();
+        int green2 = Color.red.getGreen();
+        if ((Math.abs(red1 - red2) < colorTolerance) &&
+                (Math.abs(green1 - green2) < colorTolerance) &&
+                (Math.abs(blue1 - blue2) < colorTolerance)) {
+            return true;
+        }
+        return false;
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java:61]: Could be simplified to return `<condition>;`

I wonder if this test should also verify the number of pixels of correct color as done in #10177.

File: test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+import javax.imageio.ImageIO;
+import javax.swing.JFrame;
+import javax.swing.JLabel;
+import javax.swing.SwingUtilities;
+import javax.swing.UIManager;
+
+public class TestNimbusLabel
+{
+    private static JFrame frame;
+    private static int colorTolerance = 5;
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java:47]: Let's add `final` to the constant.

File: test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java (hunk 8 above)
Comment (prsadhuk) [reply test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java]: ok

File: test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java (hunk 10 above)
Comment (prsadhuk) [reply test/jdk/javax/swing/plaf/nimbus/TestNimbusLabel.java]: ok

File: src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLabelUI.java (hunk 3 above)
Comment (aivanov-jdk) [reply src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLabelUI.java:222]: Do other L&Fs respect the `Label.foreground` color if the label text contains HTML?

Why is it set in each paint then? It should be set once in `installDefaults` and it should be removed in `uninstallDefaults`. In addition to that, it is to be installed if and only if `label.setForeground` is `null` or `instanceof UIResource`, and uninstalled if it is `instanceof UIResource`. If the user explicitly set foreground color of the label, it must not be overridden by the L&Fs.

**Review by aivanov-jdk — CHANGES_REQUESTED** [review]

File: src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLabelUI.java [hunk 11]
```diff
@@ -69,6 +71,7 @@ public static ComponentUI createUI(JComponent c){
     @Override
     protected void installDefaults(JLabel c) {
         updateStyle(c);
+        fgColor =  UIManager.getColor("Label.foreground");
```
Comment (turbanoff) [inline src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLabelUI.java:74]: let's remove redundant space after `=`

File: src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLabelUI.java (hunk 3 above)
Comment (prsadhuk) [reply src/java.desktop/share/classes/javax/swing/plaf/synth/SynthLabelUI.java]: Fair point..Updated to set in installDefaults

**Comment by mrserb** [comment]
It would be good to understand how all that properties are handled by the skin.
For example can you please check how the next property in the NimbusDefaults.java is handled by the L&F.
`addColor(d, "Label[Disabled].textForeground", "nimbusDisabledText", 0.0f, 0.0f, 0.0f, 0);`
Does it works fine? If yes probably we should add something like
`addColor(d, "Label[Enabled].textForeground", "nimbusDisabledText", 0.0f, 0.0f, 0.0f, 0);`
?

**Comment by prsadhuk** [comment]
Carried forward in https://github.com/openjdk/jdk/pull/11660

**Pull request closed without merging**